In [1]:
import numpy as np
import pandas as pd
import re


# Spécification du type de données pour la colonne 'SIREN_SIRET'
df = pd.read_csv('data/ecoles.csv', sep=';', dtype={'SIREN_SIRET': str})
pd.set_option('display.max_columns', 69)
pd.set_option('display.max_rows', 500)

In [2]:
list(df.columns)

['Identifiant_de_l_etablissement',
 'Nom_etablissement',
 'Type_etablissement',
 'Statut_public_prive',
 'Adresse_1',
 'Adresse_2',
 'Adresse_3',
 'Code postal',
 'Code_commune',
 'Nom_commune',
 'Code_departement',
 'Code_academie',
 'Code_region',
 'Ecole_maternelle',
 'Ecole_elementaire',
 'Voie_generale',
 'Voie_technologique',
 'Voie_professionnelle',
 'Telephone',
 'Fax',
 'Web',
 'Mail',
 'Restauration',
 'Hebergement',
 'ULIS',
 'Apprentissage',
 'Segpa',
 'Section_arts',
 'Section_cinema',
 'Section_theatre',
 'Section_sport',
 'Section_internationale',
 'Section_europeenne',
 'Lycee_Agricole',
 'Lycee_militaire',
 'Lycee_des_metiers',
 'Post_BAC',
 'Appartenance_Education_Prioritaire',
 'GRETA',
 'SIREN_SIRET',
 'Nombre_d_eleves',
 'Fiche_onisep',
 'position',
 'Type_contrat_prive',
 'Libelle_departement',
 'Libelle_academie',
 'Libelle_region',
 'coordonnee_X',
 'coordonnee_Y',
 'epsg',
 'nom_circonscription',
 'latitude',
 'longitude',
 'precision_localisation',
 'date_ouve

In [3]:
df.head()

Identifiant_de_l_etablissement                     Nom_etablissement  \
0                       0691513V  Ecole élémentaire De Flévieu Le Haut   
1                       0691541A      Ecole maternelle Edouard Herriot   
2                       0691579S        Ecole élémentaire Jean d'Azieu   
3                       0691707F        Ecole maternelle Pablo Picasso   
4                       0691811U                      Ecole maternelle   

  Type_etablissement Statut_public_prive                 Adresse_1 Adresse_2  \
0              Ecole              Public  FLEVIEU LE HAUT 35 R BAR       NaN   
1              Ecole              Public       2 avenue de la Gare       NaN   
2              Ecole              Public      2 rue Hector Berlioz       NaN   
3              Ecole              Public  21 avenue de Haute Roche       NaN   
4              Ecole              Public         Place de la Poste       NaN   

             Adresse_3  Code postal Code_commune    Nom_commune  \
0         69360 TERNAY        69360        69297         Ternay   
1      69800 ST PRIEST        69800        69290   Saint-Priest   
2          69740 GENAS        69740        69277          Genas   
3  69310 PIERRE BENITE        69310        69152  Pierre-Bénite   
4      69380 CHATILLON        69380        69050      Châtillon   

  Code_departement  Code_academie  Code_region  Ecole_maternelle  \
0               69             10           84               0.0   
1               69             10           84               1.0   
2               69             10           84               0.0   
3               69             10           84               1.0   
4               69             10           84               1.0   

   Ecole_elementaire  Voie_generale  Voie_technologique  Voie_professionnelle  \
0                1.0            NaN                 NaN                   NaN   
1                0.0            NaN                 NaN                   NaN   
2                1.0            NaN                 NaN                   NaN   
3                0.0            NaN                 NaN                   NaN   
4                0.0            NaN                 NaN                   NaN   

    Telephone  Fax  Web                    Mail  Restauration  Hebergement  \
0  0472246566  NaN  NaN  ce.0691513V@ac-lyon.fr           1.0          0.0   
1  0478205338  NaN  NaN  ce.0691541A@ac-lyon.fr           0.0          0.0   
2  0478901088  NaN  NaN  ce.0691579S@ac-lyon.fr           1.0          0.0   
3  0478514099  NaN  NaN  ce.0691707F@ac-lyon.fr           1.0          0.0   
4  0478439827  NaN  NaN  ce.0691811U@ac-lyon.fr           1.0          0.0   

   ULIS  Apprentissage  Segpa  Section_arts  Section_cinema  Section_theatre  \
0   0.0            NaN    NaN           NaN             NaN              NaN   
1   0.0            NaN    NaN           NaN             NaN              NaN   
2   0.0            NaN    NaN           NaN             NaN              NaN   
3   0.0            NaN    NaN           NaN             NaN              NaN   
4   0.0            NaN    NaN           NaN             NaN              NaN   

   Section_sport  Section_internationale  Section_europeenne  Lycee_Agricole  \
0            NaN                     NaN                 NaN             NaN   
1            NaN                     NaN                 NaN             NaN   
2            NaN                     NaN                 NaN             NaN   
3            NaN                     NaN                 NaN             NaN   
4            NaN                     NaN                 NaN             NaN   

   Lycee_militaire  Lycee_des_metiers  Post_BAC  \
0              NaN                NaN       NaN   
1              NaN                NaN       NaN   
2              NaN                NaN       NaN   
3              NaN                NaN       NaN   
4              NaN                NaN       NaN   

  Appartenance_Education_Prioritaire  G

#  Filtrage des champs  

In [4]:
df.drop(columns=[
         'Identifiant_de_l_etablissement',
         'Type_etablissement',
         'Adresse_2',
         'Code postal',
         'Code_commune',
         'Nom_commune',
         'Code_academie',
         'Code_region',
         'Ecole_maternelle',
         'Voie_generale',
         'Voie_technologique',
         'Voie_professionnelle',
         'Fax',
         'Web',
         'Restauration',
         'Hebergement',
         'ULIS',
         'Apprentissage',
         'Segpa',
         'Section_arts',
         'Section_cinema',
         'Section_theatre',
         'Section_sport',
         'Section_internationale',
         'Section_europeenne',
         'Lycee_Agricole',
         'Lycee_militaire',
         'Lycee_des_metiers',
         'Post_BAC',
         'Appartenance_Education_Prioritaire',
         'GRETA',
         'SIREN_SIRET',
         'Fiche_onisep',
         'position',
         'Type_contrat_prive',
         'Libelle_departement',
         'Libelle_academie',
         'Libelle_region',
         'coordonnee_X',
         'coordonnee_Y',
         'epsg',
         'nom_circonscription',
         'latitude',
         'longitude',
         'precision_localisation',
         'date_ouverture',
         'date_maj_ligne',
         'etat',
         'ministere_tutelle',
         'etablissement_multi_lignes',
         'rpi_concentre',
         'rpi_disperse',
         'code_nature',
         'libelle_nature',
         'Code_type_contrat_prive',
         'PIAL',
         'etablissement_mere',
         'type_rattachement_etablissement_mere',
         'code_bassin_formation',
         'libelle_bassin_formation'], inplace=True)

In [5]:
df.head(5)

Nom_etablissement Statut_public_prive  \
0  Ecole élémentaire De Flévieu Le Haut              Public   
1      Ecole maternelle Edouard Herriot              Public   
2        Ecole élémentaire Jean d'Azieu              Public   
3        Ecole maternelle Pablo Picasso              Public   
4                      Ecole maternelle              Public   

                  Adresse_1            Adresse_3 Code_departement  \
0  FLEVIEU LE HAUT 35 R BAR         69360 TERNAY               69   
1       2 avenue de la Gare      69800 ST PRIEST               69   
2      2 rue Hector Berlioz          69740 GENAS               69   
3  21 avenue de Haute Roche  69310 PIERRE BENITE               69   
4         Place de la Poste      69380 CHATILLON               69   

   Ecole_elementaire   Telephone                    Mail  Nombre_d_eleves  
0                1.0  0472246566  ce.0691513V@ac-lyon.fr            151.0  
1                0.0  0478205338  ce.0691541A@ac-lyon.fr            249.0  
2                1.0  0478901088  ce.0691579S@ac-lyon.fr            254.0  
3                0.0  0478514099  ce.0691707F@ac-lyon.fr            281.0  
4                0.0  0478439827  ce.0691811U@ac-lyon.fr             86.0

#  Filtrage écoles élémentaires uniquement 

In [6]:
df['Ecole_elementaire'].value_counts()

1.0    36966
0.0    13665
Name: Ecole_elementaire, dtype: int64

In [7]:
filt = (df['Ecole_elementaire'] == 1.0)

In [8]:
df = df.loc[filt]

In [9]:
df['Ecole_elementaire'].value_counts()

1.0    36966
Name: Ecole_elementaire, dtype: int64

# Filtrage des adresses mails non valides

In [10]:
df.shape

(36966, 9)

On filtre la forme des mails grâce à une regex et on élimine les ecole sans adresse mail.

In [11]:
mask = df['Mail'].str.contains(r'[^@]+@[^@]+\.[^@]+') & (df['Mail'] != np.nan)
df = df[mask]
df.shape

(36653, 9)

In [12]:
df['Mail'].str.contains(r'[^@]+@[^@]+\.[^@]+').value_counts()

True    36653
Name: Mail, dtype: int64

Il ne reste bien que des école élémentaires au nombre de 36653 (Publiques + privées), avec des emails valides.

#   Trie par départements 

In [13]:
# Trie par ordre croissant
df = df.sort_values(by='Code_departement')

# Récupération des données de df en liste


In [28]:
df['Adresse'] = df['Adresse_1'] + ' ' + df['Adresse_3']

In [106]:
df.head(50)

Name Group Membership  \
40454                 ECOLE ELEMENTAIRE PUBLIQUE               01   
56085                             Ecole primaire               01   
56087                             Ecole primaire               01   
56088                             Ecole primaire               01   
56089                             Ecole primaire               01   
56090                             Ecole primaire               01   
56091                             Ecole primaire               01   
56092                             Ecole primaire               01   
56093                             Ecole primaire               01   
56094                             Ecole primaire               01   
56095                             Ecole primaire               01   
56097                          Ecole élémentaire               01   
56098           Ecole primaire françoise convert               01   
49448                             Ecole primaire               01   
56099                             Ecole primaire               01   
56102         Ecole primaire privée Saint Joseph               01   
56103  Ecole élémentaire privée Saint Jean Bosco               01   
56106              Ecole élémentaire Les Sources               01   
56108                             Ecole primaire               01   
56109                          Ecole élémentaire               01   
56112                          Ecole élémentaire               01   
56113                  Ecole primaire René Rendu               01   
56115                Ecole élémentaire Les Lilas               01   
9775                Ecole primaire des Balmettes               01   
9777                              Ecole primaire               01   
19044              Ecole élémentaire Jean Ferrat               01   
56116                   Ecole élémentaire Centre               01   
56100            Ecole primaire  Alphonse Baudin               01   
9774                              Ecole primaire               01   
9772      Ecole primaire privée Sainte Madeleine               01   
9771          Ecole primaire privée Saint Joseph               01   
56070         Ecole primaire privée Sainte Marie               01   
56072                             Ecole primaire               01   
56073                             Ecole primaire               01   
56074                             Ecole primaire               01   
56075                             Ecole primaire               01   
56076                          Ecole élémentaire               01   
56077                          Ecole élémentaire               01   
56078                             Ecole primaire               01   
56079                             Ecole primaire               01   
56080             Ecole primaire Robert Doisneau               01   
56081                             Ecole primaire               01   
56082                          Ecole élémentaire               01   
56083                             Ecole primaire               01   
56084                          Ecole élémentaire               01   
9766          Ecole primaire privée Jeanne d'Arc               01   
9767                       Ecole primaire privée               01   
36720     ECOLE MATERNELLE PRIVEE LES TOURNESOLS               01   
5682          Ecole primaire privée Saint Joseph               01   
49454                             Ecole primaire               01   

      Phone 1 - Value        E-mail 1 - Value  \
40454      0472250117  ce.0010501P@ac-lyon.fr   
56085      0474007231  ce.0010436U@ac-lyon.fr   
56087      0474356013  ce.0010463Y@ac-lyon.fr   
56088      0474614845  ce.0010508X@ac-lyon.fr   
56089      0474397428  ce.0010530W@ac-lyon.fr   
56090      0474361265  ce.0010538E@ac-lyon.fr   
56091      0479872450  ce.0010613L@ac-lyon.fr   
56092      0479819726  ce.0010614M@ac-lyon.fr   
56093      0474399549  ce.0010647Y@ac-lyon.fr   
56094      0474240025  ce.0010670Y

On enleve les colonnes inutiles

In [31]:
df = df.drop(columns=['Statut_public_prive', 'Adresse_1', 'Adresse_3', 'Ecole_elementaire', 'Nombre_d_eleves'])

In [36]:
df.shape

(36653, 5)

In [37]:
df.head(1)

Nom_etablissement Code_departement   Telephone  \
40454  ECOLE ELEMENTAIRE PUBLIQUE               01  0472250117   

                         Mail                      Adresse  
40454  ce.0010501P@ac-lyon.fr  Grande rue 01360 BRESSOLLES

In [15]:
data = df.values.tolist()

On créer une database avec les champs du formulaire utilisateur google contact depuis nos données sous forme de liste data (plus performant de creer une dataFrame depuis une liste que de creer une dataframe vide puis de l'implementer avec les données d'une autre dataframe)

In [38]:
df.head(1)

Nom_etablissement Code_departement   Telephone  \
40454  ECOLE ELEMENTAIRE PUBLIQUE               01  0472250117   

                         Mail                      Adresse  
40454  ce.0010501P@ac-lyon.fr  Grande rue 01360 BRESSOLLES

| df      | df_contact |
| ----------- | ----------- |
| Nom_etablissement      | Name       |
| Adresse  | Address 1 - Formatted        |
| Code_departement   | Group Membership        |
| Telephone   | Phone 1 - Value        |
| Mail   | E-mail 1 - Value        |


In [ ]:
df_contacts = pd.DataFrame(data, columns=[''])

# ========= Test d'export en .csv ==========


In [21]:
df_contact

Name Given Name  Additional Name  \
0  préfixe prénonom deuxieme prénom nom suffixe   prénonom  deuxieme prénom   

  Family Name  Yomi Name    Given Name Yomi Additional Name Yomi  \
0         nom        NaN  prénom phonétique  2 prénom phonétique   

  Family Name Yomi Name Prefix Name Suffix  Initials Nickname  Short Name  \
0        nom phone     préfixe     suffixe       NaN   pseudo         NaN   

   Maiden Name  Birthday  Gender  Location  Billing Information  \
0          NaN       NaN     NaN       NaN                  NaN   

   Directory Server  Mileage  Occupation  Hobby  Sensitivity  Priority  \
0               NaN      NaN         NaN    NaN          NaN       NaN   

   Subject          Notes  Language  Photo Group Membership E-mail 1 - Type  \
0      NaN  note de texte       NaN    NaN     * myContacts          * Home   

  E-mail 1 - Value Phone 1 - Type  Phone 1 - Value  Address 1 - Type  \
0   test@gmail.com           Home          8374747               NaN   

                               Address 1 - Formatted  \
0  36 Avenue du Maréchal de Lattre de Tassigny\n9...   

                            Address 1 - Street Address 1 - City  \
0  36 Avenue du Maréchal de Lattre de Tassigny           Cachan   

   Address 1 - PO Box  Address 1 - Region  Address 1 - Postal Code  \
0                 NaN                 NaN                    94230   

  Address 1 - Country  Address 1 - Extended Address  Organization 1 - Type  \
0                  FR                           NaN                    NaN   

  Organization 1 - Name  Organization 1 - Yomi Name Organization 1 - Title  \
0            entreprise                         NaN               fonction   

   Organization 1 - Department  Organization 1 - Symbol  \
0                          NaN                      NaN   

   Organization 1 - Location  Organization 1 - Job Description  
0                        NaN                               NaN

In [22]:
df_contact = df_contact.append({
        'Name':'test_titre',
        'Address 1 - Formatted': 'test_addr',
        'Group Membership': 'test_group',
        'Phone 1 - Value': '92.43.43.43.43.43-34',
        'E-mail 1 - Value':'testemail.com',
        
    }, ignore_index=True)

In [23]:
df_contact.to_csv('data/test.csv')

# ======== Fin du test =========

# Implementation de la df_contact avec les informations traitées du .csv de l'éducation nationale

In [100]:
test = {
    'toto': [43, 34, 32],
    'tata': [44, 3, 4],
    'coco':[45, 2, 3], 
}
test2 = ['toto', 'daz', 'ds', 'coco', 'qdssqd', 'tata']
df_1 = pd.DataFrame(test)
df_2 = pd.DataFrame(columns=test2)

In [101]:
col_to_add = ['ds', 'qdssqd', 'daz']
df_1[col_to_add] = [np.nan for i in range(len(col_to_add))]
df_1

toto  tata  coco  ds  qdssqd  daz
0    43    44    45 NaN     NaN  NaN
1    34     3     2 NaN     NaN  NaN
2    32     4     3 NaN     NaN  NaN

In [102]:
df_2

Empty DataFrame
Columns: [toto, daz, ds, coco, qdssqd, tata]
Index: []

In [103]:
df_2.append(df_1)

toto  daz   ds coco qdssqd tata
0   43  NaN  NaN   45    NaN   44
1   34  NaN  NaN    2    NaN    3
2   32  NaN  NaN    3    NaN    4

A FAIRE

- changer les noms de colonnes de df pour les meme que df_contact # fait
- ajouter a df les colones manquante correspondante au champs de df_contact
- creer df_contacts depuis la liste de colonne de df_contact
- df_contacts.append(df) pour retrouver l'ordre des champs
- exporter en csv

In [90]:
df.columns = ['Name', 'Group Membership', 'Phone 1 - Value', 'E-mail 1 - Value', 'Address 1 - Formatted']

In [108]:
contact_col = list(df_contact.columns)
df_col = list(df.columns)

In [109]:
col_a_ajouter = [e for e in contact_col if e not in df_col]

In [111]:
df[col_a_ajouter] = [np.nan for i in range(len(col_a_ajouter))]

In [113]:
df_contacts = pd.DataFrame(columns=contact_col)

In [121]:
df_contacts = df_contacts.append(df)

In [122]:
df_contacts.head(1)

Name Given Name Additional Name Family Name  \
40454  ECOLE ELEMENTAIRE PUBLIQUE        NaN             NaN         NaN   

      Yomi Name Given Name Yomi Additional Name Yomi Family Name Yomi  \
40454       NaN             NaN                  NaN              NaN   

      Name Prefix Name Suffix Initials Nickname Short Name Maiden Name  \
40454         NaN         NaN      NaN      NaN        NaN         NaN   

      Birthday Gender Location Billing Information Directory Server Mileage  \
40454      NaN    NaN      NaN                 NaN              NaN     NaN   

      Occupation Hobby Sensitivity Priority Subject Notes Language Photo  \
40454        NaN   NaN         NaN      NaN     NaN   NaN      NaN   NaN   

      Group Membership E-mail 1 - Type        E-mail 1 - Value Phone 1 - Type  \
40454               01             NaN  ce.0010501P@ac-lyon.fr            NaN   

      Phone 1 - Value Address 1 - Type        Address 1 - Formatted  \
40454      0472250117              NaN  Grande rue 01360 BRESSOLLES   

      Address 1 - Street Address 1 - City Address 1 - PO Box  \
40454                NaN              NaN                NaN   

      Address 1 - Region Address 1 - Postal Code Address 1 - Country  \
40454                NaN                     NaN                 NaN   

      Address 1 - Extended Address Organization 1 - Type  \
40454                          NaN                   NaN   

      Organization 1 - Name Organization 1 - Yomi Name Organization 1 - Title  \
40454                   NaN                        NaN                    NaN   

      Organization 1 - Department Organization 1 - Symbol  \
40454                         NaN                     NaN   

      Organization 1 - Location Organization 1 - Job Description  
40454                       NaN                              NaN

In [124]:
df_contacts.shape

(36653, 50)

In [125]:
df_contacts_1 = df_contacts[:7500]
df_contacts_2 = df_contacts[7500:15000]
df_contacts_3 = df_contacts[15000:22500]
df_contacts_4 = df_contacts[22500:30000]
df_contacts_5 = df_contacts[30000:]


In [126]:
df_contacts_1.to_csv('data/df_contacts_1.csv')
df_contacts_2.to_csv('data/df_contacts_2.csv')
df_contacts_3.to_csv('data/df_contacts_3.csv')
df_contacts_4.to_csv('data/df_contacts_4.csv')
df_contacts_5.to_csv('data/df_contacts_5.csv')